clients/include/testing_hpr.hpp

```diff
@@ -0,0 +1,190 @@
+/* ************************************************************************
+ * Copyright 2018-2020 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+
+#include "cblas_interface.hpp"
+#include "flops.hpp"
+#include "near.hpp"
+#include "norm.hpp"
+#include "rocblas.hpp"
+#include "rocblas_init.hpp"
+#include "rocblas_math.hpp"
+#include "rocblas_random.hpp"
+#include "rocblas_test.hpp"
+#include "rocblas_vector.hpp"
+#include "unit.hpp"
+#include "utility.hpp"
+
+template <typename T>
+void testing_hpr_bad_arg()
+{
+    using U = decltype(std::real(T(0)));
+
+    rocblas_fill         uplo  = rocblas_fill_upper;
+    rocblas_int          N     = 100;
+    rocblas_int          incx  = 1;
+    U                    alpha = 0.6;
+    rocblas_local_handle handle;
+
+    size_t abs_incx = incx >= 0 ? incx : -incx;
+    size_t size_A   = size_t(N) * (N + 1) / 2;
+    size_t size_x   = size_t(N) * abs_incx;
+
+    // allocate memory on device
+    device_vector<T> dA_1(size_A);
+    device_vector<T> dx(size_x);
+    CHECK_HIP_ERROR(dA_1.memcheck());
+    CHECK_HIP_ERROR(dx.memcheck());
+
+    EXPECT_ROCBLAS_STATUS((rocblas_hpr<T, U>)(handle, rocblas_fill_full, N, &alpha, dx, incx, dA_1),
+                          rocblas_status_invalid_value);
+
+    EXPECT_ROCBLAS_STATUS((rocblas_hpr<T, U>)(handle, uplo, N, &alpha, nullptr, incx, dA_1),
+                          rocblas_status_invalid_pointer);
+
+    EXPECT_ROCBLAS_STATUS((rocblas_hpr<T, U>)(handle, uplo, N, &alpha, dx, incx, nullptr),
+                          rocblas_status_invalid_pointer);
+
+    EXPECT_ROCBLAS_STATUS((rocblas_hpr<T, U>)(nullptr, uplo, N, &alpha, dx, incx, dA_1),
+                          rocblas_status_invalid_handle);
+}
+
+template <typename T>
+void testing_hpr(const Arguments& arg)
+{
+    using U = decltype(std::real(T(0)));
```

Comment:
It would be nice to make this functionality available globally ... with traits ...

template < typename T >
struct precision_traits
{
  using comp_t = T;
};

template <>
struct precision_traits<rocblas_float_complex>
{
  using comp_t = float;
};

template <>
struct precision_traits<rocblas_double_complex>
{
  using comp_t = double;
};

template < typename T >
using precision_part_t = typename precision_traits< T >::comp_t;

Then you fetch your type with precision_part_t< T >.

Reply:
Added this in library/.../utility.h now. Willing to move somewhere else if people have concerns. Considered putting into rocblas-complex-types.h or in clients/.../utility.h